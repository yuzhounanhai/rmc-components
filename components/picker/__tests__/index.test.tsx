import React from 'react';
import { Simulate, act } from 'react-dom/test-utils';
import baseTest from '../../../tests/common/baseTest';
import baseReactDOMTest, { $$ } from '../../../tests/common/baseReactDOMTest';
import { sleep } from '../../../tests/common/utils';
import { mount } from 'enzyme';
import { defaultPrefixCls } from '../../_config/dict';
import Button from '../../button';
import Picker from '../index';

jest.mock('../../picker-view/index', () => (props) => (
  <div
    className="mock-components-trigger"
    onClick={() => {
      typeof props.onChange === 'function' && props.onChange(['b', 'b1']);
    }}
  >
    btn
  </div>
));

const dataCascade = [
  {
    key: 'a',
    value: 'a',
    children: [
      {
        key: 'a1',
        value: 'a1',
      },
      {
        key: 'a2',
        value: 'a2',
      },
    ],
  },
  {
    key: 'b',
    value: 'b',
    children: [
      {
        key: 'b1',
        value: 'b1',
      },
      {
        key: 'b2',
        value: 'b2',
      },
      {
        key: 'b3',
        value: 'b3',
      }
    ],
  }
]

describe('Picker', () => {
  baseReactDOMTest();

  baseTest(
    <Picker
      data={dataCascade}
    >
      btn
    </Picker>
  );

  baseTest(
    <Picker
      data={dataCascade}
    >
      btn
      <br/>
      btn
    </Picker>
  );

  baseTest(
    <Picker
      data={dataCascade}
    >
      <Button onClick={jest.fn()}>btn</Button>
    </Picker>
  );

  baseTest(
    <Picker
      data={dataCascade}
    >
      {
        (openPop, value, currentValue) => (
          <div>
            <div className="btn" onClick={openPop}>btn</div>
            <p>{`${value}-${currentValue}`}</p>
          </div>
        )
      }
    </Picker>
  );

  baseTest(
    <Picker
      data={dataCascade}
      children={null}
    />
  );

  it('should Picker render correctly.', () => {
    const onCreatePop = jest.fn();
    const wrapper1 = mount(
      <Picker
        data={dataCascade}
        onCreatePop={onCreatePop}
      >
        btn
      </Picker>
    );
    expect(wrapper1.find(`.${defaultPrefixCls}-picker-trigger`)).toHaveLength(1);
    wrapper1.find(`.${defaultPrefixCls}-picker-trigger`).simulate('click');
    expect(onCreatePop).toHaveBeenCalledTimes(1);
    const btnClickCb = jest.fn();
    const wrapper2 = mount(
      <Picker
        data={dataCascade}
        onCreatePop={onCreatePop}
      >
        <Button onClick={btnClickCb}>btn</Button>
      </Picker>
    );
    expect(wrapper2.find(`.${defaultPrefixCls}-picker-trigger`)).toHaveLength(0);
    wrapper2.find(Button).simulate('click');
    expect(onCreatePop).toHaveBeenCalledTimes(2);
    expect(btnClickCb).toHaveBeenCalled();
    const wrapper3 = mount(
      <Picker
        data={dataCascade}
        onCreatePop={onCreatePop}
      >
        {
          (openPop, value, currentValue) => (
            <div>
              <div className="btn" onClick={openPop}>btn</div>
              <p>{`${value}-${currentValue}`}</p>
            </div>
          )
        }
      </Picker>
    );
    expect(wrapper2.find(`.${defaultPrefixCls}-picker-trigger`)).toHaveLength(0);
    wrapper3.find('.btn').simulate('click');
    expect(onCreatePop).toHaveBeenCalledTimes(3);
  });
  
  it('should destory elements whick created by "Drawer" when trigger close event.', async () => {
    const wrapper = mount(
      <Picker
        data={dataCascade}
      >
        <div className="trigger-btn">btn</div>
      </Picker>
    );
    wrapper.find('.trigger-btn').simulate('click');
    expect($$(`.${defaultPrefixCls}-picker-wrapper`)).not.toBeNull();
    await act(async () => {
      await sleep(50);
    });
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    Simulate.click($$(`.${defaultPrefixCls}-picker-cancel`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    expect($$(`.${defaultPrefixCls}-picker-wrapper`)).toBeNull();
    wrapper.find('.trigger-btn').simulate('click');
    await act(async () => {
      await sleep(50);
    });
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.click($$(`.${defaultPrefixCls}-picker-ok`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    expect($$(`.${defaultPrefixCls}-picker-wrapper`)).toBeNull();

    let v1: any, v2: any;
    const wrapper1 = mount(
      <Picker
        data={dataCascade}
        defaultValue={['a', 'a1']}
        onCancel={v => {
          v1 = v;
        }}
        onOk={v => {
          v2 = v;
        }}
      >
        <div className="trigger-btn">btn</div>
      </Picker>
    );
    wrapper1.find('.trigger-btn').simulate('click');
    await act(async () => {
      await sleep(50);
    });
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    Simulate.click($$(`.mock-components-trigger`));
    Simulate.click($$(`.${defaultPrefixCls}-picker-cancel`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    ['a', 'a1'].forEach((item, idx) => {
      expect(item).toBe(v1[idx]);
    });
    wrapper1.find('.trigger-btn').simulate('click');
    await act(async () => {
      await sleep(50);
    });
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    Simulate.click($$(`.mock-components-trigger`));
    Simulate.click($$(`.${defaultPrefixCls}-picker-ok`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-slide`));
    Simulate.transitionEnd($$(`.${defaultPrefixCls}-fade`));
    ['b', 'b1'].forEach((item, idx) => {
      expect(item).toBe(v2[idx]);
    });
  });

  afterAll(() => {
    jest.unmock('../../picker-view/index');
  });
});