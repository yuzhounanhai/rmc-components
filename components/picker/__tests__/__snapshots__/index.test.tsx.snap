// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picker component snapshots tests 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "rmcc-picker-trigger",
    },
    "children": Array [
      Node {
        "data": "btn",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Picker component snapshots tests 2`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "rmcc-picker-trigger",
    },
    "children": Array [
      Node {
        "data": "btn",
        "next": Node {
          "attribs": Object {},
          "children": Array [],
          "name": "br",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Node {
            "data": "btn",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "text",
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
      Node {
        "attribs": Object {},
        "children": Array [],
        "name": "br",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "data": "btn",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "text",
        },
        "parent": [Circular],
        "prev": Node {
          "data": "btn",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
      Node {
        "data": "btn",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {},
          "children": Array [],
          "name": "br",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Node {
            "data": "btn",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "type": "text",
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Picker component snapshots tests 3`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "rmcc-button",
      "role": "button",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "rmcc-button-children",
        },
        "children": Array [
          Node {
            "data": "btn",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "a",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "role": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "role": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Picker component snapshots tests 4`] = `
initialize {
  "0": Node {
    "attribs": Object {},
    "children": Array [
      Node {
        "attribs": Object {
          "class": "btn",
        },
        "children": Array [
          Node {
            "data": "btn",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "data": "-",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "p",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Node {
        "attribs": Object {},
        "children": Array [
          Node {
            "data": "-",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "p",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "btn",
          },
          "children": Array [
            Node {
              "data": "btn",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`Picker component snapshots tests 5`] = `
Object {
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
